0002-crypto-conditions/0002-crypto-conditions.md

```diff
@@ -110,662 +116,889 @@ informative:
+Note that the example given, while useful to demonstrate the structure, has less practical security value that the use of an RSA-SHA-256 or ED25519-SHA-256 subfulfillment. Since the subfulfillment is a PREIMAGE-SHA-256, the validation of which ignores the incoming message, as long as the prefix, maxMessagelength and preimage provided in the subfulfillment are correct, the parent PREFIX-SHA-256 fulfillment will validate.
 
-threshold
-: threshold MUST be an integer in the range 1 ... 2^32 - 1. In order to fulfill a threshold condition, the weights of the provided fulfillments MUST be greater than or equal to the threshold.
+In this case, wrapping the PREIMAGE-SHA-256 crypto-condition in the PREFIX-SHA-256 crypto-condition, has the effect of enforcing a message length of 0 bytes.
 
-subconditions
-: is the set of subconditions, each provided as a tuple of weight and condition.
+Note also, any change to the PREFIX-SHA-256 crypto-condition's prefix and maxMessageLength values result in a different fingerprint value, effectively namespacing the underlying preimage and re-hashing it. The result is a new crypto-condition with a new and unique fingerprint with no change to the underlying sub-crypto-condition.
 
-weight
-: is the numeric weight of this subcondition, i.e. how many times it counts against the threshold.
+## THRESHOLD-SHA-256 {#threshold-sha-256-condition-type}
+THRESHOLD-SHA-256 is assigned the type ID 2. It relies on the availability of the SHA-256 digest algorithm and any other algorithms required by any of its sub-crypto-conditions as it is a compound crypto-condition type.
+
+### Cost {#threshold-sha-256-condition-type-cost}
 
-condition
-: is the subcondition.
+The cost is the sum of the F.threshold largest cost values of all sub-conditions, added to 1024 times the total number of sub-conditions.
 
+    cost = (sum of largest F.threshold subcondition.cost values) + 1024 * F.subconditions.count
 
-### Fulfillment {#threshold-sha-256-condition-type-fulfillment}
+For example, if a threshold crypto-condition contains 5 sub-conditions with costs of 64, 64, 82, 84 and 84 and has a threshold of 3, the cost is equal to the sum of the largest three sub-condition costs (82 + 84 + 84 = 250) plus 1024 times the number of sub-conditions (1024 * 5 = 5120): 5370
 
-    ThresholdSha256FulfillmentPayload ::= SEQUENCE {
-      threshold INTEGER (0..4294967295),
-      subfulfillments SEQUENCE OF ThresholdSubfulfillment
+### ASN.1 {#threshold-sha-256-condition-asn1}
+
+    -- Condition Fingerprint
+    ThresholdFingerprintContents ::= SEQUENCE {
+      threshold            INTEGER (1..65535),
+      subconditions        SET OF Condition
     }
 
-    ThresholdSubfulfillment ::= SEQUENCE {
-      weight INTEGER (0..4294967295) DEFAULT 1,
-      condition Condition OPTIONAL,
-      fulfillment Fulfillment OPTIONAL
+    -- Fulfillment 
+    ThresholdFulfillment ::= SEQUENCE {
+      subfulfillments      SEQUENCE OF Fulfillment,
```

Comment:
It may be misleadingly named, but there is no ambiguity. X.680 clearly says:

> **3.8.68 sequence-of types** Types defined by referencing a single component type; each value in the sequence-of type is an ordered list of zero, one or more values of the component type.
> [...]
> **3.8.73 set-of types** Types defined by referencing a single component type; each value in the set-of type is an unordered list of zero, one or more values of the component type.

A list is unambiguously something that can contain duplicate entries, so it's perfectly clear that this is allowed. Note also that the only difference in wording between `sequence-of` and `set-of` is the prefix *un*ordered, so if you're worried about sets not allowing duplicate entries, switching to a sequence wouldn't help.

X.690 says to encode (BER) a set-of exactly like a sequence-of, except that the ordering may or may not be preserved:

> **8.12 Encoding of a set-of value**
> **8.12.1** The encoding of a set-of value shall be constructed.
> **8.12.2** The text of 8.10.2 applies.
> **8.12.3** The order of data values need not be preserved by the encoding and subsequent decoding.

DER adds one more rule:

> **11.6 Set-of components**
> The encodings of the component values of a set-of value shall appear in ascending order, the encodings being compared as octet strings with the shorter components being padded at their trailing end with 0-octets.
> *NOTE – The padding octets are for comparison purposes only and do not appear in the encodings*

`SEQUENCE OF` is the **wrong** since we don't care about the order. I noticed the mistake while writing the test suite. The test suite was generating the exact same binary as the implementation-under-test in all cases except one: They would sometimes differ on the ordering of the subfulfillments and subconditions in an encoded threshold fulfillment. This is precisely why `SET OF` is in the standard. When you don't care about the ordering, `SEQUENCE OF` is a source of non-canonicality and `SET OF` fixes it by sorting the items.